phac-nml/irida-next
# Feature: Auditing

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds in DB-level logging to track changes to records using the `logidze` gem and gives us the ability to undo changes (updates) and restore a previous version of a record.

**Note: Deleting a record currently removes all data and the db level log for the record. In a future PR we will be adding in a gem `paranoia` which allows us to soft-delete records so that we can still view the changes made to the record but it will be hidden from view. This will allow us to revert any deletions as required.**

Resolves #107

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Pull this branch and run bundle
2. Run the migrations. Highly recommend dropping the database and recreating it.
3. Login and create a project
4. Open a terminal and within it open up rails console using the command `rails c`
5. Find the project using `Project.last.namespace` 
6. You should see something similar to the screenshot below

![image](https://github.com/phac-nml/irida-next/assets/25466211/8c493ff9-ab8e-4c1c-a3c7-5a9bdd226c0c)

7. Also try this out with all the other models (Group, Sample, Member, etc)
8. Make some changes to some of the records and then try step 5 and 7 again. You should see something similar to the screenshot below:

![image](https://github.com/phac-nml/irida-next/assets/25466211/4f2fa830-e458-431d-9a57-05a349997ffe)

9. Try reverting changes using `record.at(VERSION_NUMBER)`
10. Find the record in rails console and verify that the current version number is set to the version selected in step 9

To see the basic api usage visit https://www.ruby-toolbox.com/projects/logidze and go to the `Usage -> Basic API` section

To see what the log format contains you can visit this link https://www.ruby-toolbox.com/projects/logidze and go to the `Log Format` section

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
